engines/sci/graphics/controls32.cpp

```diff
@@ -353,4 +366,303 @@ void GfxControls32::flashCursor(TextEditor &editor) {
 		_nextCursorFlashTick = g_sci->getTickCount() + 30;
 	}
 }
+
+
+reg_t GfxControls32::registerScrollWindow(ScrollWindow *scrollWindow) {
+	_lastScrollWindowId += 1;
+	_scrollWindows[_lastScrollWindowId.getOffset()] = scrollWindow;
+	return _lastScrollWindowId;
+}
+
+ScrollWindow *GfxControls32::getScrollWindow(reg_t id) {
+	Common::HashMap<int, ScrollWindow *>::iterator i;
+	i = _scrollWindows.find(id.getOffset());
+	if (i != _scrollWindows.end())
+		return i->_value;
+	else
+		return nullptr;
+}
+
+void GfxControls32::deregisterScrollWindow(reg_t id) {
+	_scrollWindows.erase(id.getOffset());
+}
+
+
+
+ScrollWindow::ScrollWindow(SegManager *segMan, const Common::Rect &rect,
+                           const Common::Point &point, reg_t plane, uint8 fore,
+                           uint8 back, GuiResourceId font, TextAlign align,
+                           uint8 border)
+: _plane(plane), _foreColor(fore), _backColor(back),
+  _borderColor(border), _fontId(font), _alignment(align), _rect50(rect),
+  _point74(point) {
+
+	_gfxText32 = new GfxText32(segMan, g_sci->_gfxCache);
+
+	_firstVisibleChar = 0;
+	_lastVisibleChar = 0;
+	_topVisibleLine = 0;
+	_bottomVisibleLine = 0;
+
+	_gfxText32->setFont(_fontId);
+
+	_fontScaledWidth = _gfxText32->_scaledWidth;
+	_fontScaledHeight = _gfxText32->_scaledHeight;
+
+	_visible = false;
+
+	_rect60.left = (_rect50.left * _fontScaledWidth) / 320;
```

Comment:
320 and 200 should be using g_sci->_gfxFrameout.getCurrentBuffer().scriptWidth/scriptHeight. (I promise someday I will go through the entire code and refactor this to be less ugly)